backend/src/govsg/middlewares/govsg-transactional.middleware.ts

```diff
@@ -0,0 +1,155 @@
+          `Template with label ${whatsappTemplateLabel} not found`
+        )
+      }
+      validateParams(govsgTemplate.params, params, whatsappTemplateLabel)
+      const govsgTransactional = await GovsgMessageTransactional.create({
+        templateId: govsgTemplate.id,
+        userId: req.session?.user?.id,
+        recipient,
+        params: params ?? null,
+        status: GovsgMessageStatus.Unsent,
+      } as unknown as GovsgMessageTransactional)
+      // insert id into req.body so that subsequent middlewares can use it
+      req.body.govsgTransactionalId = govsgTransactional.id
+      next()
+    }
+
+    function validateParams(
+      govsgTemplateParams: GovsgTemplate['params'],
+      params: Record<string, string> | undefined,
+      whatsappTemplateLabel: string
+    ) {
+      // no params required or provided
+      if (
+        (!params && govsgTemplateParams === null) ||
+        govsgTemplateParams?.length === 0
+      )
+        return
+      // no params provided, but params required
+      if (
+        (!params && govsgTemplateParams !== null) ||
+        govsgTemplateParams?.length !== 0
+      ) {
+        throw new ApiValidationError(
+          `missing params for template ${whatsappTemplateLabel}`
+        )
+      }
+      // no params required, but params provided
+      // not sure whether throwing this error is necessary
+      if (params && !govsgTemplateParams) {
+        throw new ApiValidationError(
+          `template ${whatsappTemplateLabel} does not accept any params`
+        )
+      }
+      if (params && govsgTemplateParams) {
+        // todo
+        // validate params actually match the array of params accepted by GovsgTemplate
+        // check to ensure both are the same length, if not throw error
+        // for key in params, find in govsgTemplate.params
+      }
+    }
+
+    async function sendMessage(
+      req: Request<unknown, unknown, ReqBodyWithId>,
+      res: Response,
+      next: NextFunction
+    ): Promise<void> {
+      const action = 'sendMessage'
+      logger.info({ message: 'Sending GovSG transactional msg', action })
+      const {
+        govsgTransactionalId,
+        whatsapp_template_label: whatsappTemplateLabel,
+      } = req.body
+      try {
+        const govsgTransactional = await GovsgMessageTransactional.findByPk(
+          govsgTransactionalId
+        )
+        if (!govsgTransactional) {
+          // practically this will never happen unless sendMessage is called before saveMessage
+          throw new ApiNotFoundError(
+            `Unable to find id ${govsgTransactionalId} in govsg_messages_transactional`
+          )
+        }
+        const messageId = await GovsgTransactionalService.sendMessage({
+          recipient: govsgTransactional.recipient,
+          templateName: whatsappTemplateLabel,
+          params: govsgTransactional.params,
+        })
+        govsgTransactional.set('status', GovsgMessageStatus.Accepted)
```

Comment:
Shouldn't this be `SENT`? `ACCEPTED` is for after the saving step isn't it?

Reply:
so after the saving step, it's `UNSENT`. then it's `ACCEPTED` after it has been accepted by the WA API. then WA will give us a callback to say that it has been sent (see [here](https://developers.facebook.com/docs/whatsapp/on-premises/webhooks/outbound#notification-types) for the list of values).

this is similar to how it's done for tx email API at the moment